cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +47,1472 @@ using parquet::detail::ColumnChunkDesc;
+                                                      int32_t clock_rate,
+                                                      int64_t* timestamp64)
+{
+  auto nanos = int64_t{0};
+  auto days  = int64_t{0};
+  cuda::std::memcpy(&nanos, int96_value.data(), sizeof(int64_t));
+  cuda::std::memcpy(&days, int96_value.data() + sizeof(int64_t), sizeof(int32_t));
+
+  cudf::duration_D duration_days{days - 2440588};
+
+  *timestamp64 = [&]() {
+    switch (clock_rate) {
+      case 1:  // seconds
+        return cuda::std::chrono::duration_cast<cudf::duration_s>(duration_days).count() +
+               cuda::std::chrono::duration_cast<cudf::duration_s>(cudf::duration_ns{nanos}).count();
+      case 1'000:  // milliseconds
+        return cuda::std::chrono::duration_cast<cudf::duration_ms>(duration_days).count() +
+               cuda::std::chrono::duration_cast<cudf::duration_ms>(cudf::duration_ns{nanos})
+                 .count();
+      case 1'000'000:  // microseconds
+        return cuda::std::chrono::duration_cast<cudf::duration_us>(duration_days).count() +
+               cuda::std::chrono::duration_cast<cudf::duration_us>(cudf::duration_ns{nanos})
+                 .count();
+      case 1'000'000'000:  // nanoseconds
+      default:
+        return cuda::std::chrono::duration_cast<cudf::duration_ns>(duration_days).count() + nanos;
+    }
+  }();
+}
+
+/**
+ * @brief Helper function to convert a int64_t to a timestamp64
+ *
+ * @param value Value to convert
+ * @param timestamp_scale Timestamp scale
+ * @return Converted timestamp64 value
+ */
+__device__ __forceinline__ int64_t convert_to_timestamp64(int64_t const value,
+                                                          int32_t timestamp_scale)
+{
+  if (timestamp_scale < 0) {
+    // round towards negative infinity
+    int32_t const sign = (value < 0);
+    return ((value + sign) / -timestamp_scale) + sign;
+  } else {
+    return value * timestamp_scale;
+  }
+}
+
+/**
+ * @brief Query `cuco::static_set`s to evaluate (many) input (in)equality predicates
+ *
+ * @param decoded_data Span of storage for decoded values from all dictionaries
+ * @param results Span of device vector start pointers to store query results, one per predicate
+ * @param scalars Pointer to scalar device views, one per predicate
+ * @param operators Pointer to corresponding (in)equality operators, one per predicate
+ * @param set_storage Pointer to the start of the bulk storage for cuco hash sets
+ * @param set_offsets Pointer to offsets into the bulk set storage for each dictionary
+ * @param value_offsets Pointer to offsets into running sum of values in each dictionary
+ * @param total_row_groups Total number of row groups
+ * @param physical_type Parquet physical type of the column
+ */
+template <typename T>
+__global__ std::enable_if_t<not std::is_same_v<T, bool> and
+                              not(cudf::is_compound<T>() and not std::is_same_v<T, string_view>),
+                            void>
```

Comment:
Use `CUDF_ENABLE_IF` when possible or `cuda::std::enable_if` since it's device function